core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -695,11 +690,16 @@ public boolean append(Object kbase, long koff, int klen, Object vbase, long voff
       assert (vlen % 8 == 0);
       assert (longArray != null);
 
-      if (numKeys == MAX_CAPACITY
-        // The map could be reused from last spill (because of no enough memory to grow),
-        // then we don't try to grow again if hit the `growthThreshold`.
-        || !canGrowArray && numKeys > growthThreshold) {
-        return false;
+      if (numKeys >= growthThreshold) {
+        if (longArray.size() / 2 == MAX_CAPACITY) {
```

Comment:
@davies is correct; but it is a slightly unintuitive way to write the condition.

```
val currentSize = longArray.size()
val newSize = currentSize * 2
val currentKeysLen = currentSize / 2
val newKeysLen = currentKeysLen * 2

if (newKeysLen > MAX_CAPACITY) then fail.
that is  if (currentKeysLen == MAX_CAPACITY) then fail // Since we allow only power of 2's for all these values.
that is  if (longArray.size() / 2 == MAX_CAPACITY)

```

Particularly given its location (in append as opposed to grow), it serves to be a bit more confusing that expected.